### PR TITLE
internal/v5: implement publish API endpoint

### PR DESCRIPTION
**Pull request by rogpeppe**
Also make sure that it's not served from the v4 API.

### CODE REVIEW

File: internal/v4/api.go [hunk 1]
```diff
@@ -128,6 +128,8 @@ func newReqHandler() ReqHandler {
 	handlers.Id["expand-id"] = resolveId(authId(h.serveExpandId))
 	handlers.Id["archive"] = h.serveArchive(handlers.Id["archive"])
 	handlers.Id["archive/"] = resolveId(h.serveArchiveFile)
+	// Publish is a new endpoint; don't provide it in v4.
+	delete(handlers.Id, "publish")
```
Comment (frankban) [inline internal/v4/api.go:132]: Cool.

**Comment by mhilton** [comment]
:+1:

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1596/
Test PASSed.

File: internal/v5/api.go [hunk 2]
```diff
@@ -1319,6 +1320,59 @@ func (h *ReqHandler) serveAdminPromulgate(id *router.ResolvedURL, w http.Respons
...
+		return badRequestf(nil, "no channels provided")
+	}
+	for _, c := range chans {
+		if !validPublishChannels[c] {
+			return badRequestf(nil, "cannot publish to %q", c)
+		}
+	}
+
+	// Retrieve the base entity so that we can check permissions.
+	baseEntity, err := h.Cache.BaseEntity(&id.URL, nil)
```
Comment (frankban) [inline internal/v5/api.go]: Shouldn't we pass the "channelacls" field here, at least for doc purposes?

**Comment by frankban** [comment]
:+1: with one question.
Thank you!

File: internal/v5/api.go (hunk 2 above)
Comment (rogpeppe) [inline internal/v5/api.go]: done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1599/
Test PASSed.

**Pull request merged**